xarray/backends/common.py

```diff
@@ -232,11 +227,8 @@ def set_variables(self, variables, check_encoding_set,
         for vn, v in iteritems(variables):
             name = _encode_variable_name(vn)
             check = vn in check_encoding_set
-            if vn not in self.variables:
```

Comment:
Can we deprecate the variables property as part of this change? That would help avoid this in the future. In general it's a bad idea for adapter layers (like our backend classes) to hold state.

Reply:
@shoyer - done. I actually removed the the `attrs` and `dimensions` properties too since they weren't really used. I left a deprecation error in there for now but can remove it assuming all the tests pass.